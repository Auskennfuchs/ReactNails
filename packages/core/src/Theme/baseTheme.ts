import { Color } from '../Color'

interface Palette {
    [name: string]: Color
}

const palette: Palette = {
    red: Color.RGB(255, 0, 0),
}

const backgroundColors: Palette = {
    red: palette.red,
}

const textColors: Palette = {
    white: Color.white,
}

export const baseTheme = {
    palette,
    backgroundColors,
    textColors,
}

export type BaseTheme = typeof baseTheme

export interface ThemeEntry {
    [name: string]: Color | string | number | ThemeEntry
}
