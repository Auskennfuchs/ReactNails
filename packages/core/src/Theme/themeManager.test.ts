import { Color } from '..'
import { addThemeComponent, resolveTheme } from '.'
import { ThemeEntry } from './baseTheme'

describe('ThemeManager', () => {
    test('add theme component', () => {
        addThemeComponent(() => ['newThemeComponent', {
            backgroundColor: Color.black,
        }])

        const resTheme = resolveTheme({})
        expect(resTheme).toMatchSnapshot()
    })

    test('add theme component with default', () => {
        addThemeComponent(() => ['newThemeComponent', {
            backgroundColor: Color.black,
        }])

        const resTheme = resolveTheme()
        expect(resTheme).toMatchSnapshot()
    })

    test('add theme component derived', () => {
        interface BasicThemeComponent extends ThemeEntry {
            backgroundColor: Color
        }

        interface ThemeBasic extends ThemeEntry {
            newThemeComponent: BasicThemeComponent
        }

        interface DerivedThemeComponent extends BasicThemeComponent {
            textColor: Color
        }

        type ResDerivedThemeComponent = {
            derivedThemeComponent: DerivedThemeComponent
        }

        addThemeComponent<ThemeEntry, BasicThemeComponent>(() => ['newThemeComponent', {
            backgroundColor: Color.black,
        }])

        addThemeComponent<ThemeBasic, DerivedThemeComponent>((res) => ['derivedThemeComponent', {
            ...res.newThemeComponent,
            textColor: Color.RGB(255, 255, 255),
        }])

        const resTheme = resolveTheme() as ResDerivedThemeComponent
        expect(resTheme.derivedThemeComponent.backgroundColor).toStrictEqual(Color.black)
    })
})
