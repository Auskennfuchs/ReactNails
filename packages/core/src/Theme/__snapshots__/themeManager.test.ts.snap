// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeManager add theme component 1`] = `
Object {
  "backgroundColors": Object {
    "red": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 50,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 100,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "newThemeComponent": Object {
    "backgroundColor": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 0,
      "lighten": [Function],
      "luminance": [Function],
      "r": 0,
      "rgb": [Function],
      "s": 0,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "palette": Object {
    "red": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 50,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 100,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "textColors": Object {
    "white": Color {
      "a": 1,
      "b": 255,
      "contrast": [Function],
      "darken": [Function],
      "g": 255,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 100,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 0,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
}
`;

exports[`ThemeManager add theme component with default 1`] = `
Object {
  "backgroundColors": Object {
    "red": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 50,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 100,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "newThemeComponent": Object {
    "backgroundColor": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 0,
      "lighten": [Function],
      "luminance": [Function],
      "r": 0,
      "rgb": [Function],
      "s": 0,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "palette": Object {
    "red": Color {
      "a": 1,
      "b": 0,
      "contrast": [Function],
      "darken": [Function],
      "g": 0,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 50,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 100,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
  "textColors": Object {
    "white": Color {
      "a": 1,
      "b": 255,
      "contrast": [Function],
      "darken": [Function],
      "g": 255,
      "getNonAlphaColor": [Function],
      "h": 0,
      "hasAlpha": [Function],
      "hex": [Function],
      "hsl": [Function],
      "l": 100,
      "lighten": [Function],
      "luminance": [Function],
      "r": 255,
      "rgb": [Function],
      "s": 0,
      "toString": [Function],
      "wcagContrast": [Function],
    },
  },
}
`;
